src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -1864,6 +1894,592 @@ internal bool TryReadPlpBytes(ref byte[] buff, int offset, int len, out int tota
+        /// <returns>True if the connection is still alive, otherwise false</returns>
+        internal bool IsConnectionAlive(bool throwOnException)
+        {
+            Debug.Assert(_parser.Connection == null || _parser.Connection.Pool != null, "Shouldn't be calling IsConnectionAlive on non-pooled connections");
+            bool isAlive = true;
+
+            if (DateTime.UtcNow.Ticks - _lastSuccessfulIOTimer._value > CheckConnectionWindow)
+            {
+                if ((_parser == null) || ((_parser.State == TdsParserState.Broken) || (_parser.State == TdsParserState.Closed)))
+                {
+                    isAlive = false;
+                    if (throwOnException)
+                    {
+                        throw SQL.ConnectionDoomed();
+                    }
+                }
+                else if ((_pendingCallbacks > 1) || ((_parser.Connection != null) && (!_parser.Connection.IsInPool)))
+                {
+                    // This connection is currently in use, assume that the connection is 'alive'
+                    // NOTE: SNICheckConnection is not currently supported for connections that are in use
+                    Debug.Assert(true, "Call to IsConnectionAlive while connection is in use");
+                }
+                else
+                {
+                    uint error;
+
+                    RuntimeHelpers.PrepareConstrainedRegions();
+                    try
+                    {
+                        TdsParser.ReliabilitySection.Assert("unreliable call to IsConnectionAlive");  // you need to setup for a thread abort somewhere before you call this method
+
+                        SniContext = SniContext.Snix_Connect;
+
+                        error = CheckConnection();
+                        if ((error != TdsEnums.SNI_SUCCESS) && (error != TdsEnums.SNI_WAIT_TIMEOUT))
+                        {
+                            // Connection is dead
+                            SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObject.IsConnectionAlive | Info | State Object Id {0}, received error {1} on idle connection", _objectID, (int)error);
+                            isAlive = false;
+                            if (throwOnException)
+                            {
+                                // Get the error from SNI so that we can throw the correct exception
+                                AddError(_parser.ProcessSNIError(this));
+                                ThrowExceptionAndWarning();
+                            }
+                        }
+                        else
+                        {
+                            _lastSuccessfulIOTimer._value = DateTime.UtcNow.Ticks;
+                        }
+                    }
+                    finally
+                    {
+                    }
```

Comment:
If you intend to eliminate `SNIPacketRelease` method call, I don't see any reason for retaining the "try-finally" block.

Reply:
I agree, but note this change will require that the call to RuntimeHelpers.PrepareConstrainedRegions prior to it is removed too. This is what I was asking @Wraith2 and @David-Engel earlier but they didn't seem fond of the idea. Let me know how to proceed on this one.

Reply:
No objections.